src/storage/kv/store.rs

```diff
@@ -25,41 +28,54 @@ use crate::storage::{
 };
 
 /// An MVCC-based transactional key-value store.
-#[derive(Clone)]
 pub struct Store {
     pub(crate) core: Arc<Core>,
+    pub(crate) is_closed: AtomicBool,
+    stop_tx: Sender<()>,
+    task_runner_handle: Arc<RwLock<Option<JoinHandle<()>>>>,
 }
 
 impl Store {
     /// Creates a new MVCC key-value store with the given options.
     /// It creates a new core with the options and wraps it in an atomic reference counter.
     /// It returns the store.
     pub fn new(opts: Options) -> Result<Self> {
-        let core = Arc::new(Core::new(opts)?);
-        Ok(Self { core })
+        // TODO: make this channel size configurable
+        let (writes_tx, writes_rx) = bounded(10000);
+        let (stop_tx, stop_rx) = bounded(1);
+
+        let core = Arc::new(Core::new(opts, writes_tx)?);
+        let task_runner_handle = TaskRunner::new(core.clone(), writes_rx, stop_rx).spawn();
+
+        Ok(Self {
+            core,
+            stop_tx,
+            is_closed: AtomicBool::new(false),
+            task_runner_handle: Arc::new(RwLock::new(Some(task_runner_handle))),
+        })
     }
 
     /// Begins a new read-write transaction.
     /// It creates a new transaction with the core and read-write mode, and sets the read timestamp from the oracle.
     /// It returns the transaction.
-    pub fn begin(&self) -> Result<Transaction> {
+    pub async fn begin(&self) -> Result<Transaction> {
         let txn = Transaction::new(self.core.clone(), Mode::ReadWrite)?;
         Ok(txn)
     }
 
     /// Begins a new transaction with the given mode.
     /// It creates a new transaction with the core and the given mode, and sets the read timestamp from the oracle.
     /// It returns the transaction.
-    pub fn begin_with_mode(&self, mode: Mode) -> Result<Transaction> {
+    pub async fn begin_with_mode(&self, mode: Mode) -> Result<Transaction> {
         let txn = Transaction::new(self.core.clone(), mode)?;
         Ok(txn)
     }
 
     /// Executes a function in a read-only transaction.
     /// It begins a new read-only transaction and executes the function with the transaction.
     /// It returns the result of the function.
-    pub fn view(&self, f: impl FnOnce(&mut Transaction) -> Result<()>) -> Result<()> {
-        let mut txn = self.begin_with_mode(Mode::ReadOnly)?;
+    pub async fn view(&self, f: impl FnOnce(&mut Transaction) -> Result<()>) -> Result<()> {
```

Comment:
Same as begin and begin_with_mode, perhaps we don't need to make this async if I'm not missing anything?

My understanding was that we might want to do commits asynchronously so that there might be only one thread dedicated for the commit pipeline, which can make it efficient thanks to cache locality, and option to potentially better io thanks to batching WAL writes (in exchange of durability)

And we might have no motivation to make begin and view functions async? 🤔

Reply:
You are absolutely correct.

Regarding the commit being async, I had a discussion with Rushmore regarding not having the current thread blocking while waiting for disk I/O. Hence, in the commit phase, the entries are pushed into a commit pipeline, as you mentioned, and once the task is done, we are notified through a channel, and hence marked as completed.

Reply:
Same confusion as with the [async mutex](https://github.com/surrealdb/surrealkv/pull/12) PR

Reply:
I am not going to include the async mutex PR in this release of async core. Will mark this as sync

Reply:
👍 Will follow #12!